renormalizer/cv/zerot.py

```diff
@@ -27,103 +28,117 @@ class SpectraZtCV(SpectraCv):
+            method=method, procedure_cv=procedure_cv,
+            rtol=rtol, b_mps=b_mps, e0=e0, cv_mps=cv_mps,
         )
-        # procedure for ground state calculation
-        if procedure_gs is None:
-            procedure_gs = [[10, 0.4], [20, 0.2], [30, 0.1], [40, 0], [40, 0]]
-        self.procedure_gs = procedure_gs
+        
+        self.a_oper = None
+
+    def init_b_mps(self):
+        # get the right hand site vector b, Ax=b
+        # b = -eta * dipole * \psi_0
 
-    def init_oper(self):
+        # only support Holstine model 0/1 exciton manifold
         if self.spectratype == "abs":
-            self.nexciton = 0
+            nexciton = 0
             dipoletype = r"a^\dagger"
-        else:
-            self.nexciton = 1
+        elif self.spectratype == "emi":
+            nexciton = 1
             dipoletype = "a"
+        
+        # procedure for ground state calculation
+        if self.procedure_gs is None:
+            self.procedure_gs = [[10, 0.4], [20, 0.2], [30, 0.1], [40, 0], [40, 0]]
 
+        # ground state calculation
+        mps = Mps.random(
+            self.mol_list, nexciton, self.procedure_gs[0][0], percent=1.0)
+        mps.optimize_config = OptimizeConfig(procedure=self.procedure_gs)
+        mps.optimize_config.method = "2site"
+        
+        energies = solver.optimize_mps_dmrg(mps, self.h_mpo)
+        e0 = min(energies)
+        
         dipole_mpo = \
             Mpo.onsite(
                 self.mol_list, dipoletype, dipole=True
             )
-        mps, self.mpo = \
-            construct_mps_mpo_2(
-                self.mol_list, self.procedure_gs[0][0], self.nexciton
-            )
-        # ground state calculation
-        mps.optimize_config = OptimizeConfig(procedure=self.procedure_gs)
-        mps.optimize_config.method = "2site"
-        self.lowest_e = optimize_mps(mps, self.mpo)
-        ket_mps = dipole_mpo.apply(mps, canonicalise=True)
-        self.b_oper = ket_mps.scale(-self.eta)
-
-    def init_mps(self):
-        self.cv_mps = Mps.random(
-            self.mol_list, 1-self.nexciton, self.m_max, percent=1.0)
+        b_mps = dipole_mpo.apply(mps.scale(-self.eta))
+        
+        return b_mps, e0
+            
+    def init_cv_mps(self):
+        # random guess of cv_mps with same qn as b_mps
+        assert self.b_mps is not None
+        # initialize guess of cv_mps
+        cv_mps = Mps.random(
+            self.mol_list, self.b_mps.qntot, self.m_max, percent=1.0)
+        logger.info(f"cv_mps random guess qntot: {cv_mps.qntot}")
+        
+        return cv_mps
 
     def oper_prepare(self, omega):
-        h_mpo = copy.deepcopy(self.mpo)
-        for ibra in range(self.mpo.pbond_list[0]):
-            h_mpo[0][0, ibra, ibra, 0] -= (self.lowest_e + omega)
-        self.a_oper = h_mpo
-
+        # set up a_oper = (H_0 - e0 - omega)
+        identity = Mpo.identity(self.mol_list).scale(-self.e0-omega)
+        self.a_oper = self.h_mpo.add(identity)
+    
```

Comment:
This change is intended to avoid doing `deepcopy` of `self.mpo`. But it also brings problems. Firstly, It will increase the bond dim thus the complexity. Secondly, how to make sure the `qntot` of `self.h_mpo` and `identity` the same so that they can add? I noticed that qntot is set to be `0` for `Mpo.identity`.

Reply:
Add an MPO with M=1 is not a big problem. The key is that `h_mpo[0][0, ibra, ibra, 0] -= (self.lowest_e + omega)` is not correct in MolList2 because the the first element in MPO[0] is not `H_0` anymore as the MolList1 implementation.
MPO addition does not need that the two operators have s name quantum number.
I think there is no problem that the `qntot` of identity operator is 0.

Reply:
it will check the equality of qn.
```
def add(self, other: "MatrixProduct"):
        assert self.qntot == other.qntot
        assert self.site_num == other.site_num
        assert self.qnidx == other.qnidx
```

Reply:
Sorry. I'm wrong. The `qntot` should be the same in addition but not necessary in multiplication.

Reply:
should a line `identity.qn_tot = self.h_mpo.qn_tot` be added here?

Reply:
> should a line `identity.qn_tot = self.h_mpo.qn_tot` be added here?

I think qntot of Hamiltonian and Identity operator should always be zero.
Thus, I think it is ok here.

Reply:
I am OK with rest of codes.